File: pump/eopatch/src/main/java/info/nightscout/androidaps/plugins/pump/eopatch/ble/task/StartBondTask.java
package info.nightscout.androidaps.plugins.pump.eopatch.ble.task;

import static info.nightscout.androidaps.plugins.pump.eopatch.core.api.StartBonding.OPTION_NUMERIC;

import android.bluetooth.BluetoothDevice;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

import info.nightscout.androidaps.plugins.pump.eopatch.core.api.StartBonding;
import info.nightscout.rx.logging.LTag;
import io.reactivex.rxjava3.core.Single;

@Singleton
public class StartBondTask extends TaskBase {
    private final StartBonding START_BOND;

    @Inject
    public StartBondTask() {
        super(TaskFunc.START_BOND);
        START_BOND = new StartBonding();
    }

    public Single<Boolean> start(String mac) {
        prefSetMacAddress(mac);
        patch.updateMacAddress(mac, false);

        return isReady()
                .concatMapSingle(v -> START_BOND.start(OPTION_NUMERIC))
                .doOnNext(this::checkResponse)
                .concatMap(response -> patch.observeBondState())
                .doOnNext(state -> {
                    if(state == BluetoothDevice.BOND_NONE) throw new Exception();
                })
                .filter(result -> result == BluetoothDevice.BOND_BONDED)
                .map(result -> true)
                .timeout(35, TimeUnit.SECONDS)
                .doOnNext(v -> prefSetMacAddress(mac))
                .doOnError(e -> {
                    prefSetMacAddress("");
                })
                .firstOrError();
    }

    private synchronized void prefSetMacAddress(String mac) {
        pm.getPatchConfig().setMacAddress(mac);
    }
}

File: pump/eopatch/src/main/kotlin/app/aaps/pump/eopatch/ble/task/StartBondTask.kt
package app.aaps.pump.eopatch.ble.task

import android.bluetooth.BluetoothDevice
import app.aaps.pump.eopatch.core.api.StartBonding
import app.aaps.pump.eopatch.core.response.BondingResponse
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.functions.Consumer
import io.reactivex.rxjava3.functions.Function
import io.reactivex.rxjava3.functions.Predicate
import java.lang.Exception
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Suppress("PrivatePropertyName")
@Singleton
class StartBondTask @Inject constructor() : TaskBase(TaskFunc.START_BOND) {

    private val START_BOND: StartBonding = StartBonding()

    fun start(mac: String): Single<Boolean> {
        prefSetMacAddress(mac)
        patch.updateMacAddress(mac, false)

        return isReady()
            .concatMapSingle<BondingResponse>(Function { START_BOND.start(StartBonding.OPTION_NUMERIC) })
            .doOnNext(Consumer { response: BondingResponse -> this.checkResponse(response) })
            .concatMap<Int>(Function { patch.observeBondState() })
            .doOnNext(Consumer { state: Int ->
                if (state == BluetoothDevice.BOND_NONE) throw Exception()
            })
            .filter(Predicate { result: Int -> result == BluetoothDevice.BOND_BONDED })
            .map<Boolean>(Function { true })
            .timeout(35, TimeUnit.SECONDS)
            .doOnNext(Consumer { prefSetMacAddress(mac) })
            .doOnError(Consumer { prefSetMacAddress("") })
            .firstOrError()
    }

    @Synchronized private fun prefSetMacAddress(mac: String) {
        patchConfig.macAddress = mac
    }
}



